### PR TITLE
BUG: fix a ValueError raised when calling table.join with large tables

### LINKED ISSUE #14167: Join tables triggers ValueError: negative dimensions are not allowed
### Description

I have 2 big QTables (7895 rows the first and 390103 the second) that I am trying to join from a series of events.

It appears that I trigger a ValueError only when using a particular file at a particular event, so the issue might be like triggered by something within my data.

I am stuck because I do not know how to debug it more that what I already did which is
- making sure it was happening there and not in many other similar cases
- comparing pairs of events from other files and checking e.g. dtypes, units, etc...

The single tables that fail joining look OK to me.

There are column names in common and the issue gets triggered with any join type.

The traceback is similar to the following,

```python
ValueError                                Traceback (most recent call last)
File /var/folders/2z/142033n17rbfy969s6h4hymw0000gn/T/ipykernel_2009/3533429509.py:1
----> 1 join(wh, phe, join_type="left")

File ~/Applications/mambaforge/envs/swgo-plot/lib/python3.9/site-packages/astropy/table/operations.py:392, in join(left, right, keys, join_type, keys_left, keys_right, uniq_col_name, table_names, metadata_conflicts, join_funcs)
    389     right = Table(right)
    391 col_name_map = OrderedDict()
--> 392 out = _join(left, right, keys, join_type,
    393             uniq_col_name, table_names, col_name_map, metadata_conflicts,
    394             join_funcs,
    395             keys_left=keys_left, keys_right=keys_right)
    397 # Merge the column and table meta data. Table subclasses might override
    398 # these methods for custom merge behavior.
    399 _merge_table_meta(out, [left, right], metadata_conflicts=metadata_conflicts)

File ~/Applications/mambaforge/envs/swgo-plot/lib/python3.9/site-packages/astropy/table/operations.py:1191, in _join(left, right, keys, join_type, uniq_col_name, table_names, col_name_map, metadata_conflicts, join_funcs, keys_left, keys_right)
   1186 # Main inner loop in Cython to compute the cartesian product
   1187 # indices for the given join type
   1188 int_join_type = {'inner': 0, 'outer': 1, 'left': 2, 'right': 3,
   1189                  'cartesian': 1}[join_type]
   1190 masked, n_out, left_out, left_mask, right_out, right_mask = \
-> 1191     _np_utils.join_inner(idxs, idx_sort, len_left, int_join_type)
   1193 out = _get_out_class([left, right])()
   1195 for out_name, dtype, shape in out_descrs:

File astropy/table/_np_utils.pyx:77, in astropy.table._np_utils.join_inner()

ValueError: negative dimensions are not allowed
```

Did any of you ever encounter this error when joining tables?

### Steps to Reproduce

the line which triggers the issue is a simple join like

```python
join(table_A, table_B)
```

with any join type.

### System Details

macOS-11.7.1-x86_64-i386-64bit
Python 3.9.13 | packaged by conda-forge | (main, May 27 2022, 17:01:00) 
[Clang 13.0.1 ]
Numpy 1.23.4
pyerfa 2.0.0.1
astropy 5.1.1
Scipy 1.9.3
Matplotlib 3.6.2

### PR DESCRIPTION
**Pull request by neutrinoceros**
### Description
Fixes #14167.
See [my comment therin](https://github.com/astropy/astropy/issues/14167#issuecomment-1855907737) for details

ping @hamogu @taldcroft

- [ ] By checking this box, the PR author has requested that maintainers do **NOT** use the "Squash and Merge" button. Maintainers should respect this when possible; however, the final decision is at the discretion of the maintainer that merges the PR.

### CODE REVIEW

File: astropy/table/_np_utils.pyx [hunk 1]
```diff
@@ -16,7 +16,7 @@ cimport numpy as np
 np.import_array()
 
 DTYPE = int
-ctypedef np.intp_t DTYPE_t
+ctypedef np.int64_t DTYPE_t
```
Comment (saimn) [inline astropy/table/_np_utils.pyx:19]: `intp` is correct I think, it's already an int64 on 64-bit cpus, similar to `Py_ssize_t`.
https://numpy.org/devdocs/reference/c-api/dtype.html#c.npy_intp

File: astropy/table/_np_utils.pyx [hunk 2]
```diff
@@ -28,8 +28,8 @@ def join_inner(np.ndarray[DTYPE_t, ndim=1] idxs,
     Do the inner-loop cartesian product for np_utils.join() processing.
     (The "inner" is about the inner loop, not inner join).
     """
-    cdef int n_out = 0
-    cdef int max_key_idxs = 0
+    cdef DTYPE_t n_out = 0
+    cdef DTYPE_t max_key_idxs = 0
```
Comment (saimn) [inline astropy/table/_np_utils.pyx:32]: Good catch.

File: astropy/table/_np_utils.pyx (hunk 1 above)
Comment (neutrinoceros) [reply astropy/table/_np_utils.pyx:19]: Oh, I think I misread https://numpy.org/doc/stable/release/1.25.0-notes.html#cython-long-t-and-ulong-t-removed and thought its use was discouraged.

Comment (neutrinoceros) [reply astropy/table/_np_utils.pyx]: done

**Comment by neutrinoceros** [comment]
@saimn do you want to have another look ?
ping @taldcroft and @hamogu for second review

**Comment by taldcroft** [comment]
@neutrinoceros - is there a convenient way to test this? Will trying to do a ridiculous `join` cause an immediate and catchable memory error, or wreak more havoc? At the least it would be good to document in the top Description a reproducible functional test that demonstrates the patch has the intended effect.

**Comment by taldcroft** [comment]
About the code itself, I defer to @saimn, I don't have a clue about C types.

**Comment by neutrinoceros** [comment]
> is there a convenient way to test this?

There is, actually ! Initially I didn't include the test because the exact exception is a private class from numpy and the test is by construction slugish, but on second thought the exception class derives from the builtin `MemoryError` so that's something we can easily write a test for without relying too much on numpy's implementation details, and I've learned of `@pytest.mark.slow`.
Let me push the test now !

**Comment by taldcroft** [comment]
The test seems more complicated than necessary. Don't we just need two tables with a single long column?

I was just testing this out on my machine and found a different exception:
```
In [18]: c = np.arange(1000000)

In [19]: t1 = Table([c])

In [20]: t2 = Table([c])

In [21]: tj = table.join(t1, t2, join_type="cartesian")
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
Cell In[21], line 1
----> 1 tj = table.join(t1, t2, join_type="cartesian")

File ~/miniconda3/envs/ska3/lib/python3.10/site-packages/astropy/table/operations.py:413, in join(left, right, keys, join_type, keys_left, keys_right, uniq_col_name, table_names, metadata_conflicts, join_funcs)
    410     right = Table(right)
    412 col_name_map = OrderedDict()
--> 413 out = _join(
    414     left,
    415     right,
    416     keys,
    417     join_type,
    418     uniq_col_name,
    419     table_names,
    420     col_name_map,
    421     metadata_conflicts,
    422     join_funcs,
    423     keys_left=keys_left,
    424     keys_right=keys_right,
    425 )
    427 # Merge the column and table meta data. Table subclasses might override
    428 # these methods for custom merge behavior.
    429 _merge_table_meta(out, [left, right], metadata_conflicts=metadata_conflicts)

File ~/miniconda3/envs/ska3/lib/python3.10/site-packages/astropy/table/operations.py:1247, in _join(left, right, keys, join_type, uniq_col_name, table_names, col_name_map, metadata_conflicts, join_funcs, keys_left, keys_right)
   1242 # Main inner loop in Cython to compute the cartesian product
   1243 # indices for the given join type
   1244 int_join_type = {"inner": 0, "outer": 1, "left": 2, "right": 3, "cartesian": 1}[
   1245     join_type
   1246 ]
-> 1247 masked, n_out, left_out, left_mask, right_out, right_mask = _np_utils.join_inner(
   1248     idxs, idx_sort, len_left, int_join_type
   1249 )
   1251 out = _get_out_class([left, right])()
   1253 for out_name, dtype, shape in out_descrs:

File astropy/table/_np_utils.pyx:78, in astropy.table._np_utils.join_inner()

ValueError: negative dimensions are not allowed
```

**Comment by neutrinoceros** [comment]
> I was just testing this out on my machine with a smaller value and found a different exception:

@taldcroft did you recompile astropy before testing ?

> I tried with 10_000_000 and this just started chewing away memory in a way that made me worry my machine would grind to a halt in a bad way. We may not want this in tests...

Indeed. This is what I meant by "sluggish". I can't think of another way to test a memory error. Maybe I could somehow mock a very limited amount of memory, but that wouldn't make the test any simpler (provided it's even feasible). Let me research that for a minute.

**Comment by hamogu** [comment]
> Maybe I could somehow mock a very limited amount of memory

I would advise against that. That sounds like a recipe for breaking when running tests on exotic hardware, e.g. in Debian testing. I'd rather not have a test (and a good long explanation including a code example on how to test this locally in the code).

**Comment by neutrinoceros** [comment]
> That sounds like a recipe for breaking when running tests on exotic hardware

Agreed. In anycase I didn't find a good way to do it even on mainstream platforms.
I'm game to replace the test with docs, but where should it leave ?

**Comment by taldcroft** [comment]
I'm with @hamogu here of not having a unit test. My original question was about testing was to find a "convenient way". From what I've seen there is no easy way, so I'd be satisfied with a one-time functional test that is documented in the PR description.

**Comment by neutrinoceros** [comment]
The PR description already links to such a script in the issue. Should I copy it here ?

**Review by taldcroft — APPROVED** [review]
Looks good, thanks!

**Pull request merged**